File: Code/Java/Top_K_Frequent_Elements.java
/*
The K most frequent elements can be found if we add all the array elements into the HashMap, and then count the frequency for each element.
Now we'll get the max frequency now initialize an array of ArrayList. here index is frequency, value is list of numbers and then add most frequent elements to the result
 */

import java.util.*;

public class Top_K_Frequent_Elements {
	
	public static int[] takeInput() {
		Scanner s= new Scanner(System.in);
		int n= s.nextInt();
		int arr[]= new int[n];
		
		for( int i=0; i< arr.length; i++) {
			arr[i]= s.nextInt();
		}
	
		return arr;
	}

	public static List<Integer> topKFrequent(int[] nums) {
		Scanner s= new Scanner(System.in);
		 int k= s.nextInt();
		 
	    //count the frequency for each element
	    HashMap<Integer, Integer> map = new HashMap<>();
	    for(int num: nums){
	        map.put(num, map.getOrDefault(num, 0) + 1);
	    }
	 
	    //get the max frequency
	    int max = 0;
	    for(Map.Entry<Integer, Integer> entry: map.entrySet()){
	        max = Math.max(max, entry.getValue());
	    }
	 
	    //initialize an array of ArrayList. index is frequency, value is list of numbers
	    ArrayList<Integer>[] arr = (ArrayList<Integer>[]) new ArrayList[max+1];
	    for(int i=1; i<=max; i++){
	        arr[i]=new ArrayList<Integer>();
	    }
	 
	    for(Map.Entry<Integer, Integer> entry: map.entrySet()){
	        int count = entry.getValue();
	        int number = entry.getKey();
	        arr[count].add(number);
	    }
	 
	    List<Integer> result = new ArrayList<Integer>();
	 
	    //add most frequent numbers to result
	    for(int j=max; j>=1; j--){
	        if(arr[j].size()>0){
	            for(int a: arr[j]){
	                result.add(a);
	                //if size==k, stop
	                if(result.size()==k){
	                    return result;
	                }
	            }
	        }
	    }
	    return result;
	}

	public static void main(String[] args) {
 
		int arr[]= takeInput();
		 List<Integer> result= topKFrequent(arr);
		System.out.println(result);
	}
}

/*
Sample Input 1:-
 arr[]= {1}
 k= 1
Sample Output 1:-
 [1]
 
Sample Input 2:-
 arr[]= {4,5,5,2,1,1,1,5,2,3,3,5,2}
 k= 3
Sample Output 2:
 [1,5,2]
 
Time Complexity:  O(n)
Space Complexity: O(n)
 */

File: Code/C++/top_k_frequent_elements.cpp
/*
    Top K Frequent Elements
    * The following code has been implemented using unordered map and priority queue 
    * The unordered map is used for storing the frequency of each element
    * The priority queue is used to store the element along with their frequencies in pairs
    * The priority queue is implemented as a Min Heap to get the Top K Frequent elements
*/

#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

vector<int> topKFrequent(vector<int>& nums, int k) 
{
    vector <int> result;
    unordered_map<int, int> freq_map;

    for(size_t i = 0; i < nums.size(); ++i)
    {
        freq_map[nums[i]]++;    
        //increasing the stored frequency of the element by 1 when found
    }

    priority_queue< pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>> > min_heap;

    for(auto it = freq_map.begin(); it != freq_map.end(); ++it)
    {
        min_heap.push({it->second, it->first});     
        //pushing the frequency and the element in pairs respectively
        if(min_heap.size() > k)
        {
            min_heap.pop();    
            //popping the element if the size of the priority queue exceeds k
        }
    }

    for(size_t i = 0; i < k; i++)
    {
        int x = min_heap.top().second;    
        //getting the element from the top 
        result.push_back(x);    
        //pushing the element into the new vector
        min_heap.pop();
    }
    return result;
}


int main()
{
    int k, n;
    cout << "Enter size of array: ";
    cin >> n;
    cout << "Enter k: ";
    cin >> k;
    vector<int> nums(n);
    cout << "Enter elements in the array: ";
    for(size_t i = 0;i < n; ++i)
    {
        cin>>nums[i];
    }

    vector<int> ans = topKFrequent(nums, k);

    cout << "Top "<< k << " Frequent elements: ";

    for(size_t i = 0;i < k;++i)
    {
        cout << ans[i] << " ";
    }
    cout << "\n";
    return 0;
}

/*
    Test Cases:
    * Case 1:
        Enter size of array: 6
        Enter k: 2
        Enter elements in the array: 1 1 1 3 3 2
        Top 2 Frequent elements: 3 1
    * Case 2:
        Enter size of array: 8
        Enter k: 2
        Enter elements in the array: 2 2 3 3 1 1 1 1
        Top 2 Frequent elements: 3 1


    Time Complexity : O(k log d + d)
    Space Complexity : O(d)
    where d is the number of unique elements in the vector

*/          
